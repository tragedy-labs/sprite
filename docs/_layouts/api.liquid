<h6>Note</h6>
<p>
  This documentation contains some AI generations. Please be aware that there could be technical inaccuracies, and/or
  hallucinations in some of the information. You are cordially invited to make corrections to the documentation as you
  see fit. Please see the contribution guide for details.
</p>