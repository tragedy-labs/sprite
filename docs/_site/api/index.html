<!DOCTYPE html><html class="default" lang="en"><head><meta charSet="utf-8"/><meta http-equiv="x-ua-compatible" content="IE=edge"/><title>@tragedy-labs/sprite</title><meta name="description" content="Documentation for @tragedy-labs/sprite"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="stylesheet" href="assets/style.css"/><link rel="stylesheet" href="assets/highlight.css"/><script defer src="assets/main.js"></script><script async src="assets/icons.js" id="tsd-icons-script"></script><script async src="assets/search.js" id="tsd-search-script"></script><script async src="assets/navigation.js" id="tsd-nav-script"></script></head><body><script>document.documentElement.dataset.theme = localStorage.getItem("tsd-theme") || "os";document.body.style.display="none";setTimeout(() => app?app.showPage():document.body.style.removeProperty("display"),500)</script><header class="tsd-page-toolbar"><div class="tsd-toolbar-contents container"><div class="table-cell" id="tsd-search" data-base="."><div class="field"><label for="tsd-search-field" class="tsd-widget tsd-toolbar-icon search no-caption"><svg width="16" height="16" viewBox="0 0 16 16" fill="none"><use href="assets/icons.svg#icon-search"></use></svg></label><input type="text" id="tsd-search-field" aria-label="Search"/></div><div class="field"><div id="tsd-toolbar-links"></div></div><ul class="results"><li class="state loading">Preparing search index...</li><li class="state failure">The search index is not available</li></ul><a href="index.html" class="title">@tragedy-labs/sprite</a></div><div class="table-cell" id="tsd-widgets"><a href="#" class="tsd-widget tsd-toolbar-icon menu no-caption" data-toggle="menu" aria-label="Menu"><svg width="16" height="16" viewBox="0 0 16 16" fill="none"><use href="assets/icons.svg#icon-menu"></use></svg></a></div></div></header><div class="container container-main"><div class="col-content"><div class="tsd-page-title"><h2>@tragedy-labs/sprite</h2></div><div class="tsd-panel tsd-typography"><a id="md:sprite" class="tsd-anchor"></a><h1><a href="#md:sprite">Sprite</a></h1><p><img src="https://github.com/tragedy-labs/sprite/actions/workflows/commit.yml/badge.svg?event=push" alt="Test workflow badge"></p>
<p>Sprite is a TypeScript driver for ArcadeDB.</p>
<p><a href="https://sprite.tragedy.dev">Read the documentation</a>, or see the examples below.</p>
<a id="md:installation" class="tsd-anchor"></a><h2><a href="#md:installation">Installation</a></h2><p>(Use whichever package manager you prefer)</p>
<p><code>&lt;npm|pnpm|yarn|bun&gt; &lt;install|add&gt; @tragedy-labs/sprite</code></p>
<a id="md:examples" class="tsd-anchor"></a><h2><a href="#md:examples">Examples</a></h2><a id="md:spriteserver" class="tsd-anchor"></a><h3><a href="#md:spriteserver">SpriteServer</a></h3><pre><code class="language-ts"><span class="hl-0">@</span><span class="hl-1">import</span><span class="hl-0"> { SpriteServer } from </span><span class="hl-2">&#39;</span><span class="hl-3">@tragedy-labs/sprite</span><span class="hl-2">&#39;</span><span class="hl-0">;</span><br/><br/><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">server</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">new</span><span class="hl-0"> </span><span class="hl-6">SpriteServer</span><span class="hl-0">({</span><br/><span class="hl-0">  username</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">aUser</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">  password</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">aPassword</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">  address</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">http://localhost:2480</span><span class="hl-2">&#39;</span><br/><span class="hl-0">});</span><br/><br/><span class="hl-4">async</span><span class="hl-0"> </span><span class="hl-4">function</span><span class="hl-0"> </span><span class="hl-6">example</span><span class="hl-0">() {</span><br/><br/><span class="hl-0">  </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">ready</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">await</span><span class="hl-0"> server.</span><span class="hl-6">serverReady</span><span class="hl-0">();</span><br/><span class="hl-0">  console.</span><span class="hl-6">log</span><span class="hl-0">(ready);</span><br/><span class="hl-0">  </span><span class="hl-8">// true;</span><br/><br/><span class="hl-0">  </span><span class="hl-4">try</span><span class="hl-0"> {</span><br/><span class="hl-0">    </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">db</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">await</span><span class="hl-0"> server.</span><span class="hl-6">createDatabase</span><span class="hl-0">(</span><span class="hl-2">&#39;</span><span class="hl-3">aDatabase</span><span class="hl-2">&#39;</span><span class="hl-0">);</span><br/><span class="hl-0">    console.</span><span class="hl-6">log</span><span class="hl-0">(db.name);</span><br/><span class="hl-0">    </span><span class="hl-8">// &#39;aDatabase&#39;</span><br/><br/><span class="hl-0">    </span><span class="hl-4">await</span><span class="hl-0"> db.</span><span class="hl-6">command</span><span class="hl-0">(</span><span class="hl-2">&#39;</span><span class="hl-3">sql</span><span class="hl-2">&#39;</span><span class="hl-0">, </span><span class="hl-2">&#39;</span><span class="hl-3">CREATE document TYPE aType</span><span class="hl-2">&#39;</span><span class="hl-0">);</span><br/><span class="hl-0">    console.</span><span class="hl-6">log</span><span class="hl-0">(result);</span><br/><span class="hl-0">    </span><span class="hl-8">// [</span><br/><span class="hl-0">    </span><span class="hl-8">//   {</span><br/><span class="hl-0">    </span><span class="hl-8">//     operation: &#39;create document type&#39;,</span><br/><span class="hl-0">    </span><span class="hl-8">//     typeName: &#39;aType&#39;</span><br/><span class="hl-0">    </span><span class="hl-8">//   }</span><br/><span class="hl-0">    </span><span class="hl-8">// ]</span><br/><br/><span class="hl-0">    </span><span class="hl-8">// CRUD must be transactional</span><br/><span class="hl-0">    </span><span class="hl-8">// Parameterized commands are used to prevent</span><br/><span class="hl-0">    </span><span class="hl-8">// SQL injection, parameters are passed in an</span><br/><span class="hl-0">    </span><span class="hl-8">// object as the third argument</span><br/><span class="hl-0">    db.</span><span class="hl-6">transaction</span><span class="hl-0">(</span><span class="hl-4">async</span><span class="hl-0"> (</span><span class="hl-9">trx</span><span class="hl-0">) </span><span class="hl-4">=&gt;</span><span class="hl-0"> {</span><br/><span class="hl-0">      </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">doc</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> trx.</span><span class="hl-6">crud</span><span class="hl-0">(</span><br/><span class="hl-0">        </span><span class="hl-2">&#39;</span><span class="hl-3">sql</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">        </span><span class="hl-3">`INSERT INTO aType SET aProperty = :aValue`</span><span class="hl-0">,</span><br/><span class="hl-0">        { aValue</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-10">1</span><span class="hl-0"> }</span><br/><span class="hl-0">      );</span><br/><span class="hl-0">    });</span><br/><span class="hl-0">  } </span><span class="hl-4">catch</span><span class="hl-0"> (error) {</span><br/><span class="hl-0">    </span><span class="hl-4">throw</span><span class="hl-0"> </span><span class="hl-4">new</span><span class="hl-0"> </span><span class="hl-6">Error</span><span class="hl-0">(</span><span class="hl-2">&#39;</span><span class="hl-3">Could not create database</span><span class="hl-2">&#39;</span><span class="hl-0">, { cause</span><span class="hl-7">:</span><span class="hl-0"> error });</span><br/><span class="hl-0">  }</span><br/><span class="hl-0">}</span><br/><br/><span class="hl-6">example</span><span class="hl-0">();</span><br/>
</code><button>Copy</button></pre>
<a id="md:spritedatabase" class="tsd-anchor"></a><h3><a href="#md:spritedatabase">SpriteDatabase</a></h3><pre><code class="language-ts"><span class="hl-0">@</span><span class="hl-1">import</span><span class="hl-0"> { SpriteDatabase } from </span><span class="hl-2">&#39;</span><span class="hl-3">@tragedy-labs/sprite</span><span class="hl-2">&#39;</span><span class="hl-0">;</span><br/><span class="hl-0">@</span><span class="hl-1">import</span><span class="hl-0"> type {</span><br/><span class="hl-0">  CreateDocumentType,</span><br/><span class="hl-0">  InsertDocument,</span><br/><span class="hl-0">  SelectFrom</span><br/><span class="hl-0">} from </span><span class="hl-2">&#39;</span><span class="hl-3">@tragedy-labs/sprite</span><span class="hl-2">&#39;</span><span class="hl-0">;</span><br/><br/><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">db</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">new</span><span class="hl-0"> </span><span class="hl-6">SpriteDatabase</span><span class="hl-0">({</span><br/><span class="hl-0">  username</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">aUser</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">  password</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">aPassword</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">  address</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">http://localhost:2480</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">  databaseName</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-2">&#39;</span><span class="hl-3">aDatabase</span><span class="hl-2">&#39;</span><br/><span class="hl-0">});</span><br/><br/><span class="hl-4">type</span><span class="hl-0"> </span><span class="hl-11">ExampleDoc</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> {</span><br/><span class="hl-0">  </span><span class="hl-12">aValue</span><span class="hl-4">:</span><span class="hl-0"> </span><span class="hl-13">string</span><span class="hl-0">;</span><br/><span class="hl-0">};</span><br/><br/><span class="hl-4">async</span><span class="hl-0"> </span><span class="hl-4">function</span><span class="hl-0"> </span><span class="hl-6">example</span><span class="hl-0">() {</span><br/><span class="hl-0">  </span><span class="hl-4">try</span><span class="hl-0"> {</span><br/><span class="hl-0">    </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">created</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">await</span><span class="hl-0"> db.</span><span class="hl-6">command</span><span class="hl-0">&lt;</span><span class="hl-14">CreateDocumentType</span><span class="hl-0">&gt;(</span><br/><span class="hl-0">      </span><span class="hl-2">&#39;</span><span class="hl-3">sql</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">      </span><span class="hl-2">&#39;</span><span class="hl-3">CREATE document TYPE aType</span><span class="hl-2">&#39;</span><br/><span class="hl-0">    );</span><br/><br/><span class="hl-0">    console.</span><span class="hl-6">log</span><span class="hl-0">(created);</span><br/><span class="hl-0">    </span><span class="hl-8">// [</span><br/><span class="hl-0">    </span><span class="hl-8">//   {</span><br/><span class="hl-0">    </span><span class="hl-8">//     operation: &#39;create document type&#39;,</span><br/><span class="hl-0">    </span><span class="hl-8">//     type: &#39;aType&#39;</span><br/><span class="hl-0">    </span><span class="hl-8">//   }</span><br/><span class="hl-0">    </span><span class="hl-8">// ]</span><br/><br/><span class="hl-0">    </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">schema</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">await</span><span class="hl-0"> db.</span><span class="hl-6">getSchema</span><span class="hl-0">();</span><br/><span class="hl-0">    console.</span><span class="hl-6">log</span><span class="hl-0">(schema);</span><br/><span class="hl-0">    </span><span class="hl-8">// [...]</span><br/><br/><span class="hl-0">    </span><span class="hl-8">// CRUD must be part of a transaction</span><br/><span class="hl-0">    </span><span class="hl-8">// Parameterized commands are used to prevent</span><br/><span class="hl-0">    </span><span class="hl-8">// SQL injection, parameters are passed in an</span><br/><span class="hl-0">    </span><span class="hl-8">// object as the third argument.</span><br/><span class="hl-0">    db.</span><span class="hl-6">transaction</span><span class="hl-0">(</span><span class="hl-4">async</span><span class="hl-0"> (</span><span class="hl-9">trx</span><span class="hl-0">) </span><span class="hl-4">=&gt;</span><span class="hl-0"> {</span><br/><span class="hl-0">      </span><span class="hl-4">await</span><span class="hl-0"> trx.</span><span class="hl-6">crud</span><span class="hl-0">(</span><br/><span class="hl-0">        </span><span class="hl-2">&#39;</span><span class="hl-3">sql</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">        </span><span class="hl-3">`INSERT INTO aType SET aProperty = :aValue`</span><span class="hl-0">,</span><br/><span class="hl-0">        { aValue</span><span class="hl-7">:</span><span class="hl-0"> </span><span class="hl-10">1</span><span class="hl-0"> }</span><br/><span class="hl-0">      );</span><br/><span class="hl-0">    });</span><br/><br/><span class="hl-0">    </span><span class="hl-4">const</span><span class="hl-0"> </span><span class="hl-5">docs</span><span class="hl-0"> </span><span class="hl-4">=</span><span class="hl-0"> </span><span class="hl-4">await</span><span class="hl-0"> db.</span><span class="hl-6">query</span><span class="hl-0">&lt;</span><span class="hl-14">SelectFrom</span><span class="hl-0">&lt;</span><span class="hl-14">ExampleDoc</span><span class="hl-0">&gt;&gt;(</span><br/><span class="hl-0">      </span><span class="hl-2">&#39;</span><span class="hl-3">sql</span><span class="hl-2">&#39;</span><span class="hl-0">,</span><br/><span class="hl-0">      </span><span class="hl-2">&#39;</span><span class="hl-3">SELECT * FROM aType</span><span class="hl-2">&#39;</span><br/><span class="hl-0">    );</span><br/><span class="hl-0">    console.</span><span class="hl-6">log</span><span class="hl-0">(docs);</span><br/><span class="hl-0">    </span><span class="hl-8">// [</span><br/><span class="hl-0">    </span><span class="hl-8">//   {</span><br/><span class="hl-0">    </span><span class="hl-8">//     aProperty: &#39;aValue&#39;</span><br/><span class="hl-0">    </span><span class="hl-8">//   }</span><br/><span class="hl-0">    </span><span class="hl-8">// ]</span><br/><br/><span class="hl-0">  } </span><span class="hl-4">catch</span><span class="hl-0"> (error) {</span><br/><span class="hl-0">    </span><span class="hl-4">throw</span><span class="hl-0"> </span><span class="hl-4">new</span><span class="hl-0"> </span><span class="hl-6">Error</span><span class="hl-0">(</span><span class="hl-2">&#39;</span><span class="hl-3">An error occured...</span><span class="hl-2">&#39;</span><span class="hl-0">, { cause</span><span class="hl-7">:</span><span class="hl-0"> error });</span><br/><span class="hl-0">  }</span><br/><span class="hl-0">}</span><br/><br/><span class="hl-6">example</span><span class="hl-0">();</span><br/>
</code><button>Copy</button></pre>
</div></div><div class="col-sidebar"><div class="page-menu"><div class="tsd-navigation settings"><details class="tsd-index-accordion"><summary class="tsd-accordion-summary"><h3><svg width="20" height="20" viewBox="0 0 24 24" fill="none"><use href="assets/icons.svg#icon-chevronDown"></use></svg>Settings</h3></summary><div class="tsd-accordion-details"><div class="tsd-filter-visibility"><h4 class="uppercase">Member Visibility</h4><form><ul id="tsd-filter-options"><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-protected" name="protected"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Protected</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-private" name="private"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Private</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-inherited" name="inherited" checked/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Inherited</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-external" name="external"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>External</span></label></li></ul></form></div><div class="tsd-theme-toggle"><h4 class="uppercase">Theme</h4><select id="tsd-theme"><option value="os">OS</option><option value="light">Light</option><option value="dark">Dark</option></select></div></div></details></div><details open class="tsd-index-accordion tsd-page-navigation"><summary class="tsd-accordion-summary"><h3><svg width="20" height="20" viewBox="0 0 24 24" fill="none"><use href="assets/icons.svg#icon-chevronDown"></use></svg>On This Page</h3></summary><div class="tsd-accordion-details"><a href="#md:sprite"><span>Sprite</span></a><ul><li><a href="#md:installation"><span>Installation</span></a></li><li><a href="#md:examples"><span>Examples</span></a></li><li><ul><li><a href="#md:spriteserver"><span>Sprite<wbr/>Server</span></a></li><li><a href="#md:spritedatabase"><span>Sprite<wbr/>Database</span></a></li></ul></li></ul></div></details></div><div class="site-menu"><nav class="tsd-navigation"><a href="modules.html" class="current"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="assets/icons.svg#icon-1"></use></svg><span>@tragedy-labs/sprite</span></a><ul class="tsd-small-nested-navigation" id="tsd-nav-container" data-base="."><li>Loading...</li></ul></nav></div></div></div><footer><p class="tsd-generator">Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p></footer><div class="overlay"></div></body></html>