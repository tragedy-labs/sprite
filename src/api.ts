export { SpriteDatabase } from './SpriteDatabase.js';
export { SpriteServer } from './SpriteServer.js';
export { SpriteTransaction } from './SpriteTransaction.js';
export { DocumentModality } from './modes/DocumentModality.js';
export { GraphModality } from './modes/GraphModality.js';

export * from './types/index.js';
