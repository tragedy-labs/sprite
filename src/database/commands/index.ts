export { EXPLAIN } from './EXPLAIN.js';
export { SELECT_SCHEMA } from './SELECT_SCHEMA.js';
