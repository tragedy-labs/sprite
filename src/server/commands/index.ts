export { CLOSE_DATABASE } from './CLOSE_DATABASE.js';
export { CONNECT_CLUSTER } from './CONNECT_CLUSTER.js';
export { CREATE_DATABASE } from './CREATE_DATABASE.js';
export { CREATE_USER, type IArcadeCreateUser } from './CREATE_USER.js';
export { DISCONNECT_CLUSTER } from './DISCONNECT_CLUSTER.js';
export { DROP_DATABASE } from './DROP_DATABASE.js';
export { DROP_USER } from './DROP_USER.js';
export { GET_SERVER_EVENTS } from './GET_SERVER_EVENTS.js';
export { OPEN_DATABASE } from './OPEN_DATABASE.js';
export { SHUTDOWN } from './SHUTDOWN.js';
