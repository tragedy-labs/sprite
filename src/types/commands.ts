import {
  ArcadeRecordCategory,
  EdgeRecordMeta,
  RecordMeta
} from './database.js';
import { DeleteFromCount } from './operators.js';

/** Base Interface for a command response from ArcadeDB */
interface CommandResponse<T = string> {
  operation: T;
}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a command that is a `BUCKET` operation
 */
type BucketOperationResponse<T = string, R = string> = Array<
  CommandResponse<R> & {
    bucketName: T;
    bucketId: number;
  }
>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `ALTER BUCKET` command
 */
export interface AlterBucket<Name = string>
  extends BucketOperationResponse<Name, 'alter bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE BUCKET` command
 */
export interface CreateBucket<Name = string>
  extends BucketOperationResponse<Name, 'create bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DROP BUCKET` command
 */
export interface DropBucket<Name = string>
  extends BucketOperationResponse<Name, 'drop bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a command that is a `TYPE` operation
 */
type TypeOperationResponse<T = string, R = string> = Array<
  CommandResponse<R> & {
    typeName: T;
  }
>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `ALTER TYPE` command
 */
export interface AlterType<Name = string>
  extends TypeOperationResponse<Name, 'alter type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE document TYPE <Name>` command
 */
export interface CreateDocumentType<Name = string>
  extends TypeOperationResponse<Name, 'create document type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE edge TYPE <Name>` command
 */
export interface CreateEdgeType<Name = string>
  extends TypeOperationResponse<Name, 'create edge type'> {}

/**
 * An array, which contain an object returned from
 * ArcadeDB for a `CREATE vertex TYPE <Name>` command
 */
export interface CreateVertexType<Name = string>
  extends TypeOperationResponse<Name, 'create vertex type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DROP document TYPE <Name>` command
 */
export interface DropType<Name = string>
  extends TypeOperationResponse<Name, 'drop type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <document-type>` or
 * a `INSERT INTO <vertex-type>` command
 */
type InsertRecord<T, C extends ArcadeRecordCategory> = Array<T & RecordMeta<C>>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <document-type>`
 * command
 * @template Type The type of the document to be inserted
 */
export interface InsertDocument<Type = unknown>
  extends InsertRecord<Type, 'd'> {}
/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <vertex-type>`
 * command
 * @template Type The type of the vertex to be inserted
 */
export interface InsertVertex<Type = unknown> extends InsertRecord<Type, 'v'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE EDGE <edge-type>` command
 * @template TypeName - The name of the edge type
 */
export type CreateEdge<Type = unknown> = Array<Type & EdgeRecordMeta>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE VERTEX <vertex-type>` command
 * @template Type - The type of the vertex to create
 */
export type CreateVertex<Type = unknown> = InsertVertex<Type>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DELETE FROM` command. The object
 * contains one property, `count` which is an
 * integer
 */
export type DeleteFrom = Array<DeleteFromCount>;
