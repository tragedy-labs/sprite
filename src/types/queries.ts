import { ArcadeRecordCategory, RecordMeta } from './database.js';

/**
 * The `DOCUMENT`, `EDGE`, or `VERTEX` with ArcadeDB
 * meta attached. If being used in a query command,
 * expect it to be delivered as an array of the
 * described object.
 */
type ArcadeRecord<T, C extends ArcadeRecordCategory> = T & RecordMeta<C>;

/**
 * The `DOCUMENT` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `DOCUMENT` type definition (without meta)
 */
export type ArcadeDocument<T = unknown> = ArcadeRecord<T, 'd'>;

/**
 * The `EDGE` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `EDGE` type definition (without meta)
 */
export type ArcadeEdge<T = unknown> = ArcadeRecord<T, 'e'>;

/**
 * The `VERTEX` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `VERTEX` type definition (without meta)
 */
export type ArcadeVertex<T = unknown> = ArcadeRecord<T, 'v'>;
