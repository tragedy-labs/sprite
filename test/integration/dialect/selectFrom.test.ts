import { AsArcadeDocuments } from 'src/api.js';
import { testClient } from './testclient.js';
import { CreateDocumentType, DropType } from 'src/types/commands.js';
import { ArcadeDocument } from 'src/types/queries.js';

const typeName = 'SelectFromTestType';

interface SelectFromTestType {
  aProperty: string;
}

interface DocumentTypes {
  [typeName]: SelectFromTestType;
}

const data: SelectFromTestType = {
  aProperty: 'aValue'
};

const db = testClient.database;

describe('SqlDialect.selectFrom()', () => {
  beforeAll(async () => {
    await db.command<CreateDocumentType<typeof typeName>>(
      'sql',
      `CREATE document TYPE ${typeName}`
    );
  });
  afterAll(async () => {
    await db.command<DropType<typeof typeName>>('sql', `DROP TYPE ${typeName}`);
  });
  it(`should select record`, async () => {
    /* Arrange */

    await db.command(
      'sql',
      `INSERT INTO ${typeName} CONTENT ${JSON.stringify([data, data])}`
    );

    /* Act */
    const records = await testClient.selectFrom<
      AsArcadeDocuments<DocumentTypes>,
      typeof typeName,
      'aProperty'
    >(typeName, {
      where: ['aProperty', '==', 'aValue']
    });

    /* Assert */
    records.forEach((record: ArcadeDocument<SelectFromTestType>) => {
      expect(record).toHaveProperty('@rid'); // Ensure each record has the '@rid' property
      expect(record['@type']).toBe(typeName); // Ensure each record has the '@type' property
      expect(record).toHaveProperty('aProperty'); // Ensure each record has the 'name' property
      expect(record.aProperty).toBe('aValue'); // Ensure 'aProperty' is 'aValue'
    });
  });

  it(`should propagate errors from the database`, async () => {
    /* Arrange, Act, & Assert */
    await expect(
      // @ts-expect-error - intentionally passing invalid type
      testClient.selectFrom('INVALID_TYPE_NAME')
    ).rejects.toMatchSnapshot();
  });
});
