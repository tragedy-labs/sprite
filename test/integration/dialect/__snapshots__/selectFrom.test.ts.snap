// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlDialect.selectFrom() should propagate errors from the database 1`] = `[Error: Could not execute select command on type:INVALID_TYPE_NAME in database SpriteIntegrationTesting]`;
